blog/DS-1311/RDS-Encryption-At-Rest.md

```diff
@@ -0,0 +1,391 @@
+	newReplica, err := s.ModifyInstance(name, *groupName, copyFrom.FilterVPCSecurityGroups(Active))
+	if err != nil {
+		return err
+	}
+	return s.cloneReplica(master, copyFrom, newReplica, binlogRetention, rootPass)
+}
+```
+
+which first calls `ModifyInstance` [routine](https://github.com/InVisionApp/ds-blog/blob/master/blog/DS-1311/code/modify_instance.go) we went over earlier and then `cloneReplica`:
+
+```go
+func (s *SDK) cloneReplica(master, copyFrom, newReplica Instance, binlogRetention int, rootPass string) error {
+	// for now only support mysql
+	if newReplica.Engine != "mysql" {
+		return nil
+	}
+
+	rc := &mysqlReplicaClone{
+		copyFrom:     copyFrom,
+		copyTo:       newReplica,
+		rootPass:     rootPass,
+		binlogRetHrs: aws.Int(binlogRetention),
+		log:          s.log,
+	}
+	dryRun := false
+	return rc.execute(s.Verbose, dryRun)
+}
+```
+where `mysqlReplicaClone.execute` does all of the heavy lifting to match DB's internals, the code for which is available in [mysql_replica_clone.go](https://github.com/InVisionApp/ds-blog/blob/master/blog/DS-1311/code/mysql_replica_clone.go).  The goal of this clone process is to take a list of DB users from source replica instance and compare it with the target and re-create any missing users and their grants.  Here's how this works:
+
+- after making [connections](https://github.com/InVisionApp/ds-blog/blob/master/blog/DS-1311/code/mysql_replica_clone.go#L94-L102) to source and target instances, we gather the users and their grants on both source and target instances using `dumpQuery` function [here](https://github.com/InVisionApp/ds-blog/blob/master/blog/DS-1311/code/mysql_replica_clone.go#L104-L119)
+- [dumpQuery](https://github.com/InVisionApp/ds-blog/blob/master/blog/DS-1311/code/mysql_replica_clone.go#L246-L324)'s job is to turn MySQL table rows into Go's `map` with it's **key** comprised of *concatenated primary key values* and it's **value** being the entire row represented as `map` with *column names as keys and column values as pointer values*.  For example, suppose the table is:
+
+```
+PK1 	PK2 	COL1	COL2
+-----	-----	-----	-----
+p1A 	p2A 	c1A 	c2A
+p1B 	p2B 	c1B 	c2B
+p1C 	p2C 	c1C 	NULL
+```
+
+then the resulting Go's map structure is as follows:
+
+```
+[p1A|p2A]{
+  [PK1]*p1A
+  [PK2]*p2A
+  [COL1]*c1A
+  [COL2]*c2A
+}
+[p1B|p2B]{
+  [PK1]*p1B
+  [PK2]*p2B
+  [COL1]*c1B
+  [COL2]*c2B
+}
+[p1C|p2C]{
+  [PK1]*p1C
+  [PK2]*p2C
+  [COL1]*c1C
+  [COL2]*<nil>
+}
+```
+
+having this data structure allows us to lookup any row by it's primary key and then lookup each column by it's name.  This is very powerful.
+
+* at his point we have our source and target users and all of the source schema level grants, we now compare the list of users and if the user is missing on the target we re-create it and it's schema level grants (if any) [here](https://github.com/InVisionApp/ds-blog/blob/master/blog/DS-1311/code/mysql_replica_clone.go#L133-L170)
```

Comment:
```suggestion
* at this point we have our source and target users and all of the source schema level grants, we now compare the list of users and if the user is missing on the target we re-create it and its schema level grants (if any) [here](https://github.com/InVisionApp/ds-blog/blob/master/blog/DS-1311/code/mysql_replica_clone.go#L133-L170)
```

Reply:
thanks - I refactored this section and the typo is fixed as part of it